heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java

```diff
@@ -0,0 +1,384 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+import com.twitter.heron.api.generated.TopologyAPI;
+import com.twitter.heron.packing.utils.Container;
+import com.twitter.heron.spi.common.Config;
+import com.twitter.heron.spi.common.Constants;
+import com.twitter.heron.spi.common.Context;
+import com.twitter.heron.spi.packing.IPacking;
+import com.twitter.heron.spi.packing.PackingPlan;
+import com.twitter.heron.spi.utils.TopologyUtils;
+
+/**
+ * FFD packing algorithm
+ * <p>
+ * This IPacking implementation generates a PackingPlan based on the
+ * First Fit Decreasing heuristic for the binpacking problem. The algorithm attempts to minimize
+ * the amount of resources used.
+ * <p>
+ * Following semantics are guaranteed:
+ * 1. Supports heterogeneous containers. The number of containers used is determined
+ * by the algorithm and not by the config file.
+ * The size of each container cannot exceed a maximum size provided by the user
+ * <p>
+ * 2. The user provides the maximum CPU, RAM and Disk that can be used by each container through
+ * the com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_CPU_REQUESTED,
+ * com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_RAM_REQUESTED,
+ * com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_DISK_REQUESTED parameters.
+ * If the parameters are not specified then a default value is used for the maximum container
+ * size.
+ * <p>
+ * 3. The ram required for one instance is calculated as:
+ * value in com.twitter.heron.api.Config.TOPOLOGY_COMPONENT_RAMMAP if exists, otherwise,
+ * the default ram value for one instance.
+ * <p>
+ * 4. The cpu required for one instance is calculated as the default cpu value for one instance.
+ * <p>
+ * 5. The disk required for one instance is calculated as the default disk value for one instance.
+ * <p>
+ * 6. The ram required for a container is calculated as:
+ * (ram for instances in container) + (ram padding for heron internal process)
+ * <p>
+ * 7. The cpu required for a container is calculated as:
+ * (cpu for instances in container) + (cpu padding for heron internal process)
+ * <p>
+ * 8. The disk required for a container is calculated as:
+ * (disk for instances in container) + (disk padding for heron internal process)
+ * <p>
+ * 9. The size of resources required by the whole topology is equal to
+ * ((# containers used by FFD)* size of each container
+ * + 2 * (size of a container that includes one instance with default resource requirements)).
+ * The extra 2 is considered for Heron internal container (i.e. the one containing Scheduler
+ * and TMaster) and for the YARN application master,
+ * <p>
+ * 7. The pack() return null if PackingPlan fails to pass the safe check, for instance,
+ * the size of ram for an instance is less than the minimal required value.
+ */
+
+public class FFDPacking implements IPacking {
```

Comment:
`FFD` is a pretty esoteric term. What if we expanded it to `FirstFitDecreasing` or `BinPacking`? I feel like expanding out the name would make it more discoverable.